Centralize task creation

DougGregor opened the pull request:
Centralize and clean up task creation facilities for the concurrency runtime. This involves a number of changes:
* Route all task creation through a single, overridable function `swift_task_create_common`. We can back-patch this with the usual runtime "override" mechanism to fix bugs or back-deploy features as necessary.
* Reduce the external entry points for task creation down to three: `swift_task_create` (used for spawning most tasks), `swift_task_create_common` (used when we already have the function pointer we need), and `swift_asyncLet_start` (used to start `async let` tasks). Make the Swift task-launching entry points small, always-emit-into-client shims over `swift_task_create` so we can evolve the API without changing the ABI.
* Use a distinct set of flags for launching tasks (rather than repurposing job flags), and add in support for copying task locals, inheriting other context, enqueuing the task (including on a specified executor), and wiring up async let storage.
* Rework the SIL builtins for task creation to use the new task creation flags and always map down to `swift_task_create`.
* Make `Task` `@frozen` because it's always going to be a wrapper over a single pointer.

Implements rdar://79584250

DougGregor committed:
Store an ExecutorRef within an executor option.

ExecutorRefs are a (pointer, witness table) pair that are meant to be
passed around. We don't need to form a reference to one because
they are ABI already.

DougGregor committed:
Simplify swift_task_create_group_future_common API and rename it.

Collapse the `group` parameter of this API into the task options, and
have existing callers set up the options appropriately. The goal for
this function is to become the centralized entry point for all task
creation, with an extensible interface.

DougGregor committed:
Create a separate set of task creation flags for `swift_task_create`.

The flags that are useful for task creation are a bit different from
the flags that go on a job. Create a separate flag set for task
creation and use that in the API for `swift_task_create`. For now,
have the callers do the remapping.

DougGregor committed:
Centralize non-group task creation on swift_task_create[_f].

Introduce a builtin `createAsyncTask` that maps to `swift_task_create`,
and use that for the non-group task creation operations based on the
task-creation flags. `swift_task_create` and the thin function version
`swift_task_create_f` go through the dynamically-replaceable
`swift_task_create_common`, where all of the task creation logic is
present.

While here, move copying of task locals and the initial scheduling of
the task into `swift_task_create_common`, enabling by separate flags.

ktoso commented on include/swift/AST/Builtins.def:
```diff
@@ -805,6 +805,17 @@ BUILTIN_MISC_OPERATION(StartAsyncLet, "startAsyncLet", "", Special)
+/// Create a new asynchronous task, given flags, options, and a function to
+/// execute.
+BUILTIN_MISC_OPERATION_WITH_SILGEN(CreateAsyncTask,
+    "createAsyncTask", "", Special)
```
wohoo ❤️

ktoso commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -396,43 +397,70 @@ task_future_wait_resume_adapter(SWIFT_ASYNC_CONTEXT AsyncContext *_context) {
+    jobFlags.task_setIsChildTask(true);
+  } else {
+    jobFlags.task_setIsChildTask(taskCreateFlags.isChildTask());
+  }
```
ah yeah that's much nicer. I had initially started implementing this by asserting here but makes sense to just infer from asynclet -> must be child. 👍

ktoso commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -396,43 +397,70 @@ task_future_wait_resume_adapter(SWIFT_ASYNC_CONTEXT AsyncContext *_context) {
+      jobFlags.task_setIsGroupChildTask(true);
+      break;
+    }
+  }
```
👍

ktoso commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -649,6 +641,49 @@ getAsyncClosureEntryPointAndContextSize(void *function,
+  taskCreateFlags.setIsChildTask(jobFlags.task_isChildTask());
+  taskCreateFlags.setIsAsyncLetTask(jobFlags.task_isAsyncLetTask());
+  return taskCreateFlags.getOpaqueValue();
+}
```
add FIXME? though perhaps we'd be able to remove this still in the same PR?

DougGregor replied:
Yes, it'll go away in a subsequent commit. I should  have marked this PR as a draft.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -379,6 +379,97 @@ struct JobFlags {
+
+  /// Whether to copy thread locals from the currently-executing task into the
+  /// newly-created task.
+  var copyThreadLocals: Bool {
```
Hmm... Is this `copyThreadLocalTaskLocals` rather?

DougGregor replied:
D'oh! Yes, task locals, not thread locals.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -379,6 +379,97 @@ struct JobFlags {
+
+  /// Whether to enqueue the newly-created task on the given executor (if
+  /// specified) or the global executor.
+  var enqueueJob: Bool {
```
nice idea :) was quite annoying to deal with the always having to enqueue "on the outside" (e.g. in swift code again)...

DougGregor replied:
Yeah, I'm trying to get the Swift entry points down to "set of flags, call into the C++ code".

ktoso reviewed (approved):
Looks great so far :) Really looking forward to removing all the other entry points 💯

DougGregor committed:
Remove CreateAsyncTaskFuture and swift_task_create_future.

We no longer need these entry points.

DougGregor committed:
Switch task group child task creation over to Builtin.createAsyncTask().

Extend the behavior of `swift_task_create_common` to also encompass
adding the pending group task (when requested) and attaching it to the
group.

DougGregor committed:
Correct naming of "CopyThreadLocals" to "CopyTaskLocals"

DougGregor committed:
Remove CreateAsyncTaskGroupFuture and swift_task_create_group_future.

We've moved everything over to `CreateAsyncTask` now.

DougGregor committed:
Use withUnsafeBytes(of:) properly for getting the address of a local variable

DougGregor committed:
Use `withUnsafePointer(to:)` rather than `withUnsafeBytes(of:)`

DougGregor committed:
Remove `swift_task_create_async_let_future`.

This is a small shim over `swift_task_create`. Use that instead.

DougGregor committed:
Hollow out `swift_asyncLet_startImpl` to just a `swift_task_create` call.

Introduce a task option record to capture the async-let storage, so
that `swift_task_create` can perform the appropriate initialization.

DougGregor committed:
Eliminate `swift_task_create_f`.

It's a tiny shim over `swift_task_create_common` that is rarely used.
Switch all of the former callers over the common API.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
The macOS full test hit the playground logger issue but was otherwise good. Trying the smoke test

DougGregor committed:
Remove unnecessary debug dump

DougGregor committed:
Introduce a `createAsyncTaskInGroup` SIL builtin.

Rather than using group task options constructed from the Swift parts
of the _Concurrency library and passed through `createAsyncTask`'s
options, introduce a separate builtin that always takes a group. Move
the responsibility for creating the options structure into IRGen, so
we don't need to expose the TaskGroupTaskOptionRecord type in Swift.

DougGregor committed:
Drop task options from the `createAsyncTask` SIL builtin.

We'll keep the task options entirely as a contract between IRGen and
the concurrency runtime, so Swift code need not deal with them.

DougGregor committed:
Move priority adjustments for async task creation into the runtime.

We're not actually performing the adjustments at the moment due to an
unrelated bug, and will want to perform them within
`swift_task_create_common` based on inheritContext and the given
priority.

DougGregor committed:
Make Swift task-creation entry points always-emit-into-client.

The `swift_task_create` entry point is our general runtime ABI for
launching tasks. Make the various Swift APIs sitting on top of it
always-emit-into-client to take them out of the ABI. This reduces the
number of ABI entry points and allows us to make more ABI-compatible
changes to the Swift side.

DougGregor committed:
Drop overridable entry point for swift_asyncLet_start.

This entry point is a small shim over `swift_task_create_common`, which
is overridable. We don't need a separate override point here.

DougGregor commented:
@swift-ci please test

ktoso commented on lib/IRGen/GenBuiltin.cpp:
```diff
@@ -270,16 +270,26 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
+    llvm::Value *taskOptions =
+        (Builtin.ID == BuiltinValueKind::CreateAsyncTask)
+        ? args.claimNext()
+        : llvm::ConstantInt::get(IGF.IGM.SwiftTaskOptionRecordPtrTy, 0);
```
TIL: oh! that's good to know

ktoso commented on include/swift/ABI/TaskOptions.h:
```diff
@@ -80,14 +80,14 @@ class TaskGroupTaskOptionRecord : public TaskOptionRecord {
 /// concurrent executor, or the enclosing actor's executor.
 class ExecutorTaskOptionRecord : public TaskOptionRecord {
-  ExecutorRef *Executor;
+  ExecutorRef Executor;
```
👍

ktoso commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -2029,11 +2029,12 @@ class TaskCreateFlags : public FlagSet<size_t> {
+    Task_CopyThreadLocals                         = 10,
+    Task_InheritContext                           = 11,
+    Task_EnqueueJob                               = 12,
+    Task_AddPendingGroupTaskUnconditionally       = 13,
```
ah, excellent to put it as a flag 👍

ktoso commented on lib/IRGen/GenCall.cpp:
```diff
@@ -3861,16 +3861,14 @@ llvm::Value *irgen::emitTaskCreate(
   // If there is a task group, emit a task group option structure to contain
   // it.
+  llvm::Value *taskOptions = llvm::ConstantInt::get(
+      IGF.IGM.SwiftTaskOptionRecordPtrTy, 0);
```
I see, that's much easier... thanks :)

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -36,9 +36,12 @@ import Swift
 /// individually schedulable as jobs.  Jobs are generally not interacted
 /// with by end-users directly, unless implementing a scheduler.
 @available(SwiftStdlib 5.5, *)
+@frozen
```
⛄ ❤️

ktoso reviewed (approved):
LGTM, thanks a lot for taking this over! Learnt a lot by reading this 👍 

No concerns, all looks great 👍

DougGregor committed:
Fix priority propagation when creating new tasks

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor committed:
Fix test for 32-bit platforms

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please smoke test macOS
